### PR TITLE
DPC-1573: Use multistage builds for V2 API and V2 Attribution

### PR DESCRIPTION
**Pull request by msnook**
### Fixes [DPC-1573](https://jira.cms.gov/browse/DPC-1573)

The V2 API and V2 Attribution Docker images currently contain all of the source code.  We should ensure that the images are lean.

### Proposed Changes

- Use multistage builds so that the final image that is used in production is lean and secure.

### Change Details

- Updated V2 API `Dockerfile` to leverage multistage builds, with a lean final image.
- Updated V2 Attribution `Dockerfile` to leverage multistage builds, with a lean final image.

### Security Implications

- [ ] new software dependencies
- [ ] security controls or supporting software altered
- [ ] new data stored or transmitted
- [ ] security checklist is completed for this change
- [ ] requires more information or team discussion to evaluate security implications
- [x] no PHI/PII is affected by this change

### Acceptance Validation

The images build successfully.  Manual inspection of the running containers was performed, and ensured that the source code has been removed.

### Feedback Requested

The scope of this modification is limited to introducing multistage builds.  However, there are a few outstanding questions which might warrant additional tickets?

(1) Is port 8080 the correct port to be exposed?  
(2) Is the CI in Github Actions building, linting, and testing these V2 containers?
(3) Should we be building Swagger documentation dynamically instead of baking it into the image from a static snapshot saved in the repo?

### CODE REVIEW

**Comment by jonfulk** [comment]
(1) Is port 8080 the correct port to be exposed?
- not sure on this, but it seems to be to me
(2) Is the CI in Github Actions building, linting, and testing these V2 containers?
- We do not yet have CI in GHA for V2
(3) Should we be building Swagger documentation dynamically instead of baking it into the image from a static snapshot saved in the repo?
- At this point, we made the decision to manually update the Swagger docs for V2. That may change in the future.

File: dpc-go/dpc-attribution/docker/Dockerfiles/Dockerfile.attribution [hunk 1]
```diff
@@ -1,19 +1,19 @@
...
-# Set the Current Working Directory inside the container
 WORKDIR /app/dpc-attribution
-
-# We want to populate the module cache based on the go.{mod,sum} files.
 COPY src/ .
 COPY configs ../configs
-
-# Build the Go app
 RUN go build  -o ./bin/attribution .
+EXPOSE 8080
```
Comment (jonfulk) [inline dpc-go/dpc-attribution/docker/Dockerfiles/Dockerfile.attribution:7]: is this a duplicate of line 16?

File: dpc-go/dpc-api/docker/Dockerfiles/Dockerfile.api [hunk 2]
```diff
@@ -1,20 +1,17 @@
-FROM golang:1.15-alpine3.12
-
+FROM golang:1.15-alpine3.12 AS builder
 RUN apk add --no-cache git
-
-# Set the Current Working Directory inside the container
```
Comment (jonfulk) [inline dpc-go/dpc-api/docker/Dockerfiles/Dockerfile.api:5]: why drop all the comments?

**Review by jonfulk — DISMISSED** [review]
👍 
Will this cause fresh images to be built each time we start V2 using `make start-v2` (due to the --no-cache flag)? If so, that would actually be helpful for development.

File: dpc-go/dpc-attribution/docker/Dockerfiles/Dockerfile.attribution (hunk 1 above)
Comment (msnook) [reply dpc-go/dpc-attribution/docker/Dockerfiles/Dockerfile.attribution]: Yup!  Good catch.  Removed in the latest commit.

File: dpc-go/dpc-api/docker/Dockerfiles/Dockerfile.api (hunk 2 above)
Comment (msnook) [reply dpc-go/dpc-api/docker/Dockerfiles/Dockerfile.api:5]: Seemed to me like the comments were redundant; the code itself was self-documenting.

**Comment by msnook** [comment]
> 👍
> Will this cause fresh images to be built each time we start V2 using `make start-v2` (due to the --no-cache flag)? If so, that would actually be helpful for development.

If that behavior is desired, you'd need to add the `--no-cache` flags to the location where the build is actually happening (i.e., in the Makefile target).

**Comment by msnook** [comment]
> (1) Is port 8080 the correct port to be exposed?
> 
> * not sure on this, but it seems to be to me
>   (2) Is the CI in Github Actions building, linting, and testing these V2 containers?
> * We do not yet have CI in GHA for V2
>   (3) Should we be building Swagger documentation dynamically instead of baking it into the image from a static snapshot saved in the repo?
> * At this point, we made the decision to manually update the Swagger docs for V2. That may change in the future.

Is there a ticket for adding CI in Github Actions for V2?  It seems like that should be in place now, so regressions aren't being introduced undetected.  We'd definitely need it in place prior to deploying to AWS environments, to ensure we don't break deployments.

**Comment by jonfulk** [comment]
> > (1) Is port 8080 the correct port to be exposed?
> 
> > 
> 
> > * not sure on this, but it seems to be to me
> 
> >   (2) Is the CI in Github Actions building, linting, and testing these V2 containers?
> 
> > * We do not yet have CI in GHA for V2
> 
> >   (3) Should we be building Swagger documentation dynamically instead of baking it into the image from a static snapshot saved in the repo?
> 
> > * At this point, we made the decision to manually update the Swagger docs for V2. That may change in the future.
> 
> 
> 
> Is there a ticket for adding CI in Github Actions for V2?  It seems like that should be in place now, so regressions aren't being introduced undetected.  We'd definitely need it in place prior to deploying to AWS environments, to ensure we don't break deployments.

We have discussed the need, so there should be.

**Review by lostpurplecow — APPROVED** [review]
🚢

File: dpc-go/dpc-api/docker/Dockerfiles/Dockerfile.api [hunk 3]
```diff
@@ -1,20 +1,17 @@
...
 COPY configs ../configs
 COPY swaggerui ../swaggerui
-
-# Build the Go app
 RUN go build  -o ./bin/api .
 
-# This container exposes port 8080 to the outside world
+FROM golang:1.15-alpine3.12
+RUN apk update upgrade
+RUN apk --no-cache add ca-certificates aws-cli curl
```
Comment (orenfromberg) [inline dpc-go/dpc-api/docker/Dockerfiles/Dockerfile.api:11]: what are these dependencies for?

**Review by orenfromberg — APPROVED** [review]
LGTM, just one question to satisfy my curiosity.

File: dpc-go/dpc-api/docker/Dockerfiles/Dockerfile.api (hunk 3 above)
Comment (msnook) [reply dpc-go/dpc-api/docker/Dockerfiles/Dockerfile.api:11]: To remain consistent with the other DPC service images, the `aws-cli` and `curl` utilities are included for troubleshooting purposes.

**Pull request merged**